### PR TITLE
JOSS publication

### LINKED ISSUE #1633: Publishing a paper in JOSS
Currently, we don't have a publication about `{lintr}`, and I would propose that we write a short paper for [Journal of Open Source Software](https://joss.theoj.org/about). _JOSS_ is open source, open access, and peer reviewed, and has no publication fees. If you aren't familiar with _JOSS_, as an example, see the [paper on `{tidyverse}`](https://joss.theoj.org/papers/10.21105/joss.01686). 

If the team has no objections to this, I can work on creating the first draft in the next few months, which can then be polished further by the team. I've already published seven papers in _JOSS_ about other packages and have a fairly good idea on how to write them. But the actual submission (whenever the final draft is ready) will need to be handled by the first or primary author.

WDYT?

### LINKED ISSUE #1884: Provide an easy way to format lint messages with user-specified width
No matter where lint messages are printed (e.g. GitHub annotations, RStudio markers, Markdown in GitHub issues, etc.), currently, there is no straightforward way to print them with a certain width. This means that the user always needs to scroll horizontally to see many of the lint messages. 

``` r
options(width = 50)
library(lintr)

lint(
  text = "x <- c();",
  linters = unnecessary_concatenation_linter()
)
#> <text>:1:6: style: [unnecessary_concatenation_linter] Unneeded concatenation without arguments. Replace the "c" call by NULL or, whenever possible, vector() seeded with the correct type and/or length.
#> x <- c();
#>      ^~~
```

<sup>Created on 2023-01-03 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

A typical project that starts using linters might throw hundreds of messages, and that's a lot of scrolling. This might be mildly annoying for an able-bodied individual like me, but must be extremely challenging and disappointing for people with mobility impairment. Almost every single lint asks them to scroll to see the full lint message. 

We should anticipate this and provide a way out.

### PR DESCRIPTION
**Pull request by IndrajeetPatil**
Closes #1633 

If you want to read a PDF containing the latest changes, you can look at the artefacts uploaded by run of [the latest workflow](https://github.com/r-lib/lintr/actions/workflows/draft-pdf.yml) that renders it.

---

If you aren't familiar with _JOSS_, as an example, see the [paper on `{tidyverse}`](https://joss.theoj.org/papers/10.21105/joss.01686). 

For more, see: https://joss.readthedocs.io/en/latest/submitting.html

---

TODO:

- [x] Wait for the next release (since the draft uses `all_linters()` function).
- [x] Resolve #1884 (this will make sure that the lint messages don't overflow in the PDF output)
- [x] Decide whether to include both examples that do and don't lint, or only the former. Since the lint message suggests the alternatives that won't lint anyway, I don't see the need to repeat examples.
- [x] Decide on the authors and the order of the authors.
- [ ] Acknowledge funding sources, if relevant

### CODE REVIEW

**Comment by IndrajeetPatil** [comment]
@jimhester, @MichaelChirico, @AshesITR I have put together a rough first draft and can already use some feedback from you on whether the draft seems like a good first step.

You can also read the [rendered PDF.](https://github.com/r-lib/lintr/blob/joss_paper/paper/paper.pdf)

Feel free to make edits. Also, we can skip CI to save some energy, since there are no code-related changes in this PR. We can run the workflows once the PR is ready for a merge.

Thanks.

**Comment by jimhester** [comment]
I think what you have is a great start, thanks for writing it up!

Do papers in JOSS typically try to measure impact in any way? e.g. through number of reverse imports / suggests, GitHub references or CRAN downloads?

I also wonder if we should mention the age / maturity of the package in some way, to contrast it with other packages that are maybe not as seasoned.

**Comment by IndrajeetPatil** [comment]
> I think what you have is a great start, thanks for writing it up!
> 
> Do papers in JOSS typically try to measure impact in any way? e.g. through number of reverse imports / suggests, GitHub references or CRAN downloads?
> 
> I also wonder if we should mention the age / maturity of the package in some way, to contrast it with other packages that are maybe not as seasoned.

Thanks for the comment, Jim. I think it makes sense to highlight these aspects. Here is a paragraph I have added to the "Benefits" section. WDYT?

> Finally, `{lintr}` has had a large and active user community since its birth in 2014 which has contributed to its rapid development, maintenance, and adoption. At the time of writing, `{lintr}` is in a mature and stable state and therefore provides a reliable API that is unlikely to feature any breaking changes.

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:37]: if this audience may not be familiar with R, it's important to clarify that R is dynamically-typed, since that colors so much of what we can/can't do with {lintr}.

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:41]:
```suggestion
As of this writing, `{lintr}` offers 85 linters
```

That's roughly on par with the # I see in e.g. `flake8`, I'm not sure "more than N!" is the right framing :)

File: paper/paper.Rmd
Comment (MichaelChirico) [reply paper/paper.Rmd:37]: So far, I don't see any mention that a primary purpose of {lintr} is helping to enforce the tidyverse style guide in particular. I think this is a useful contextual touchstone.

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:57]: If the audience may not know R, it will be good to discuss the difference of vector/scalar operators a bit more in the context of R as a language.

Comment (IndrajeetPatil) [reply paper/paper.Rmd:59]: Hmm, I am not sure if we need to get into that. 

Will it be better to use some other linter as an example to avoid having to explain vector/scalar operators?

File: paper/paper.Rmd
Comment (IndrajeetPatil) [inline paper/paper.Rmd:82]: These code chunks are currently unevaluated because I am not sure if we want to include examples that don't lint. 

If we decide not to, these chunks can be removed.

File: paper/paper.Rmd
Comment (IndrajeetPatil) [inline paper/paper.Rmd:5]: I have added only a preliminary list of authors. I am not sure about what should either the details (authors can do this themselves) or the order here.

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd]: > So far, I don't see any mention that a primary purpose of {lintr} is helping to enforce the tidyverse style guide in particular. I think this is a useful contextual touchstone.

@MichaelChirico It's a bit ironic that (AFAICT) none of the tidyverse packages actually use `{lintr}` in their development workflow 😅

File: .Rbuildignore
Comment (MichaelChirico) [inline .Rbuildignore:31]: Nit: just `^paper$`? that will exclude the whole directory. At a glance I thought we were adding new top-level files instead of a new directory.

File: .github/workflows/draft-pdf.yml
Comment (MichaelChirico) [inline .github/workflows/draft-pdf.yml:4]: Does this mean this draft will get rebuilt on every push/pull request? Might be better to just do this on merge to `main`?

Comment (MichaelChirico) [reply .github/workflows/draft-pdf.yml:4]: (PS in general, do we need to include `master` here? since we only really use `main`...)

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:53]: > can be run from the command line

This might make it seem like one could do `./lintr file --linters=assignment_linter()` or something like that, but we don't ship any command-line binary, you have to invoke `Rscript` (right?)

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:71]:
```suggestion
`{lintr}` offers linters that can detect problematic antipatterns and suggest alternatives that follow best practices.
```

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:84]: is this meant to have no `yes`/`no` arguments?

File: paper/paper.Rmd
Comment (MichaelChirico) [reply paper/paper.Rmd:82]: I lean towards dropping them.

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:91]:
```suggestion
Sometimes the users might not be aware of a more efficient way offered by R for carrying out a computation. `{lintr}` offers linters to improve code efficiency by avoiding common inefficient patterns.
```

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:100]: Here, the "no lint" example does add value, but maybe better served by explaining in text why `any(is.na())` is deemed inefficient.

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:109]:
```suggestion
Coders spend significantly more time reading than writing code [@mcconnell2004code]. Thus, writing readable code makes the code more maintainable and reduces the possibility of introducing bugs stemming from a poor understanding of the code.
```

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:114]: I would suggest a more compelling readability linter for the paper, maybe `comparison_negation_linter()`?

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:129]: > https://style.tidyverse.org/index.html

Also linked above, should it be a bibliography entry?

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:133]: Maybe `object_name_linter()` would be a better choice? This one is kind-of a {magrittr} bug, harder to think of it as being a "core" aspect of the tidyverse guide. Whereas tidyverse naming style does "stick out" a bit in R which has naming style wild west.

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:127]: I think it'd be good to include a "common mistakes" example too. That's a good example of when {lintr} giving early feedback is important (usually these issues would be caught by tests, but (1) that happens "much" later than typing the code (2) assumes high test coverage).

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:145]: One thing I'd also highlight is our design for extensibility -- yes, we enforce the Tidyverse guide by default, but we designed the package to fit many different style guides for any person/team/org that wants to deviate or even to start their own guide.

File: .github/workflows/draft-pdf.yml
Comment (IndrajeetPatil) [reply .github/workflows/draft-pdf.yml:4]: I had just copied this file from JOSS. But we can get rid of `master` here without any issues.

We can delete this workflow once the paper is published. I have added a TODO to that effect.

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd]: True. I no longer mention the command line here.

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd]: Created a bib entry. Currently treating as a manual since it's not a published book.

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd:126]: How about this?

> `{lintr}` is designed for extensibility by allowing users to easily create custom linting rules, enabling flexible and tailored code analysis to suit specific coding conventions, style guides, and project requirements.

**Comment by IndrajeetPatil** [comment]
Do we also want to include a sticker in the paper, like in [the tidyverse paper](https://joss.theoj.org/papers/10.21105/joss.01686)?

If yes, we will need to resolve #1188 first.

**Comment by IndrajeetPatil** [comment]
@MichaelChirico Do you want to give this another read?

I think I have already addressed all of your concerns from the previous round of review.

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:11]:
```suggestion
    orcid: 0000-0003-0787-087X
```

**Review by MichaelChirico — COMMENTED** [review]
Thanks! We're getting close :)

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd]: It is an overkill, but let's do it anyway, lest it's marked as a lint in future by some doclinter 😉

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd]: I agree that this is a better example. I have included it in the paper.

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd:190]: I agree with the first point, but I think the second point/example would be too technical for this type of article (since it involves many concepts that we never discuss here, e.g. XPath) that is supposed to target a wide range of readers. WDYT?

Comment (MichaelChirico) [reply paper/paper.Rmd:190]: Yea, I had a similar worry. How about instead of a specific example, a paragraph at a high enough level about the ability to write custom linters?

Comment (IndrajeetPatil) [reply paper/paper.Rmd]: I have added a new section to discuss this. Can you please have a look?

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:53]: Should we add a citation about "static code analysis", e.g. just the Wikipedia page?

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:117]: Blanking on the best phrasing here, but I think it's worth calling out that `!x==2` is even more unreadable than `!(x==2)`.

```suggestion
Note also the complicated operator precedence. The more readable alternative here uses `!=`:
```

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:102]: Below, we don't offer the "bad"/"good" pairing of snippets, should we remove this "good" example here?

File: paper/paper.Rmd
Comment (MichaelChirico) [inline paper/paper.Rmd:189]:
```suggestion
  - Create new linters (by leveraging functions like `lintr::make_linter_from_xpath()`) tailored to match project- or organization-specific coding standards.
```

File: paper/paper.Rmd
Comment (MichaelChirico) [reply paper/paper.Rmd:53]: Similarly, does it make sense to mention popular linters in other languages? I'm not sure about how general the context we should provide for JOSS is.

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd:102]: But this chunk is a companion to `any_is_na_linter_with_lint` chunk, which comes right before it.

Comment (MichaelChirico) [reply paper/paper.Rmd:102]: Right... there's still some inconsistency. We have companion chunks for `comparison_negation_linter()` and `any_is_na_linter()`, but not `object_name_linter()` or `redundant_ifelse_linter()`. Since the corresponding "good" code for `redundant_ifelse_linter()` is covered in the text, maybe it's only `object_name_linter()` that's the outlier? Or do you think it's fine as is?

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd]: There is a comprehensive summary of tools here: https://github.com/analysis-tools-dev/static-analysis

But I can't think of a good place where to place it, or how to cite it.

**Comment by MichaelChirico** [comment]
GTG I think. I don't have a good sense of the standards JOSS is aiming for so not sure when the marginal benefit from improving will drop off. Thanks so much for driving this!!

cc @jimhester / @AshesITR for any other high-level comments.

File: paper/paper.Rmd
Comment (IndrajeetPatil) [reply paper/paper.Rmd:109]: Good point. I hadn't included the no-lint versions for these because they felt too trivial, but it does feel inconsistent to not see them. I have included them now.

**Comment by IndrajeetPatil** [comment]
> GTG I think. I don't have a good sense of the standards JOSS is aiming for so not sure when the marginal benefit from improving will drop off. Thanks so much for driving this!!

@MichaelChirico I think good examples to give you a feel for their standards, you can have a look at these papers (combined estimated reading time ~ 30 min): [[1](https://joss.theoj.org/papers/10.21105/joss.01686)], [[2](https://joss.theoj.org/papers/10.21105/joss.03167)], [[3](https://joss.theoj.org/papers/10.21105/joss.03139)]

> cc @jimhester / @AshesITR for any other high-level comments.

@jimhester, @AshesITR: If you want to read a PDF containing the latest changes, you can look at the artefacts uploaded by run of [the latest workflow](https://github.com/r-lib/lintr/actions/workflows/draft-pdf.yml) that renders it. (I've also updated my top-level comment with this info.)

**Comment by IndrajeetPatil** [comment]
@MichaelChirico While we await feedback on the manuscript from others, we can probably work on resolving this item in the meantime:

>  Decide on the authors and the order of the authors.

**Comment by MichaelChirico** [comment]
For authors, since order apparently matters (in my field we just do alphabetical), what's natural to me is (1) Jim first (2) any other author in DESCRIPTION next, alphabetically.

**Comment by AshesITR** [comment]
I'm fine with any order. Regarding the list of authors my suggestion would be to stick to who is listed in DESCRIPTION.

**Comment by IndrajeetPatil** [comment]
@jimhester, @russHyde, @fangly, @renkun-ken Can you please check your affiliations and also update the affiliation and ORCID (if available)?

Also, if you have time, it would be great to get some feedback on the draft of the paper :)

(Please see the top-level comment in this PR for context)

File: paper/paper.md
Comment (jimhester) [inline paper/paper.md:8]:
```suggestion
    orcid: 0000-0002-2739-7082
```

**Review by jimhester — APPROVED** [review]

**Pull request merged**
